### PR TITLE
Hash to g1

### PR DESCRIPTION
**Pull request by nano-o**
Still needs some cleanup but should work. A few assumptions that could be proved also remain.

### CODE REVIEW

File: proof/hash_to_g1.saw [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+] simp_then_admit;
+
+unsquare_rewrite <- simp_then_admit {{ \a -> Fp.sq a == Fp.mul (a,a) }};
+
+left_fold_mul <- simp_then_admit {{ \x y z -> Fp.mul (x, Fp.mul (y,z)) == Fp.mul (Fp.mul (x, y),z) }};
+
+// this one is not included in algebra_simpset below:
+right_fold_mul <- simp_then_admit {{ \x y z -> Fp.mul (Fp.mul (x, y),z) == Fp.mul (x, Fp.mul (y,z)) }};
+
+simp_div_rewrites <- for [
```
Comment (msaaltink) [inline proof/hash_to_g1.saw:192]: I don't think any of these are true when x is 0.

Comment (nano-o) [reply proof/hash_to_g1.saw:193]: Yes, I need to fix that

File: spec/Serialization.cry [hunk 2]
```diff
@@ -17,9 +17,11 @@ import BLSFieldExtras (sqrt_fp, is_square_fp, sqrt_fp2, is_square_fp2)
 
 //// Serialization (with compression)
 
+// TODO: draft 7 changed this
```
Comment (msaaltink) [inline spec/Serialization.cry:20]: Do we have a plan for this TODO?  Is it that the code version we analyse is compliant with an earlier version of the spec that the v9 we used?

File: spec/implementation/HashToG1.cry [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    zzs = [Fp.sq Z, Fp.sq (Fp.sq Z)]
+      # (join`{parts=inf} [[Fp.mul (zi0, zi1), Fp.sq zi1] | zi0 <- zzs | zi1 <- tail zzs]) // TODO: why is parts=inf necessary here?
+
+map_fp_times_Zz : {m1,m2,n} (fin n, fin m2, m1 >= n, m2 >= n) => [m1]t_Fp -> [m2]t_Fp -> [n]t_Fp
+map_fp_times_Zz ks zzs = zipWith (curry Fp.mul) (take`{front=n}ks) (drop`{back=n}zzs)
+
+map_fp kzzs x = last rs where
+  rs = [last kzzs] # [Fp.add (Fp.mul (ri, x), kzzi) | kzzi <- tail (reverse kzzs) | ri <- rs]
+
+to_jacobian xn xd yn yd = (X, Y, Z) where
```
Comment (msaaltink) [inline spec/implementation/HashToG1.cry:96]: cleanup note for later: we have other versions of this, `projective_pair_to_jacobian_imp` in `psi.saw` and `G2SubGroupCheck`. Those two for Fp_2, so not in conflict with this definitions, but perhaps worth collecting together.

File: proof/hash_to_g1.saw [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+// we also need this:
+let recip_sqrt_fp_3mod4_spec = do {
+    out_ptr <- crucible_alloc vec384_type;
+    (inp, inp_ptr) <- ptr_to_fresh_readonly "inp" vec384_type;
+    crucible_precond {{ fp_invariant inp }};
+    crucible_execute_func [out_ptr, inp_ptr];
+    crucible_points_to out_ptr (crucible_term {{fp_rep (Fp.div (Fp.field_unit, sqrt_fp (fp_abs inp))) }});
+    };
+
+sgn0x_pty_mont_384_ov <- admit "sgn0x_pty_mont_384" sgn0x_pty_mont_384_spec;
```
Comment (msaaltink) [inline proof/hash_to_g1.saw:108]: I thought you had proved this override

File: proof/hash_to_g1.saw [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+
+// rewrite rules to get rid of the Aux version of some definitions
+
+aux_rewrites <- for [
+    {{ Aux::A' == A' }}
+  , {{ Aux::Z == Z }}
+  , {{ Aux::B' == B' }}
+  , {{ \x -> Aux::sgn0 x == sgn0 x}}
+  , {{ \x -> Aux::sqrt x == sqrt x}}
+] simp_then_admit;
```
Comment (msaaltink) [inline proof/hash_to_g1.saw:161]: Are these not trivially provable?

File: proof/hash_to_g1.saw [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+swu_on_curve <- simp_then_admit {{ \u -> is_point_affine Curve_E' (map_to_curve_simple_swu u) == True }};
+
+// point_dadd implements add:
+dadd_thm <- admit_cryptol (rewrite (cryptol_ss ()) {{ \ p1 p2 -> affinify Curve_E' (point_dadd Fp p1 p2 A' False) == add Curve_E' (affinify Curve_E' p1) (affinify Curve_E' p2) }});
+
+// the addition of two points on the curve is on the curve:
+point_dadd_on_curve <- simp_then_admit {{ \p q -> is_point_projective Curve_E' (point_dadd Fp p q A' False) == if is_point_projective Curve_E' p /\ is_point_projective Curve_E' q then True else apply is_point_projective Curve_E' (point_dadd Fp p q A' False) }};
+
+// adding a point to its double can be done with POINTonE1_add (cheaper than dadd) because the double of a point is never equal to the point unless the point is point_O:
+// TODO: is this rule okay?
```
Comment (msaaltink) [inline proof/hash_to_g1.saw:181]: This looks OK.  I can prove all the cases using the assumptions in `ec_mult.saw`, so it is no worse than those.

File: proof/hash_to_g1.saw (hunk 4 above)
Comment (nano-o) [reply proof/hash_to_g1.saw:108]: Yes indeed

File: proof/hash_to_g1.saw (hunk 5 above)
Comment (nano-o) [reply proof/hash_to_g1.saw:161]: Yes

File: proof/hash_to_g1.saw (hunk 4 above)
Comment (nano-o) [reply proof/hash_to_g1.saw:108]: Actually no, it's only been tested with quickcheck.

File: spec/Serialization.cry (hunk 2 above)
Comment (nano-o) [reply spec/Serialization.cry:20]: The code is compliant with the spec, but the spec is weird: there are two versions of sign, used at different places. The code does both at once (see sgn0.saw). At the moment I forgot which is used where.

Comment (nano-o) [reply spec/Serialization.cry:20]: I checked and the spec says serialization should be done using the ZCash format, so what's in Serialization.cry is correct.

**Pull request merged**